api/datasets.go

```diff
@@ -597,24 +580,67 @@ func (h DatasetHandlers) bodyHandler(w http.ResponseWriter, r *http.Request) {
 	// if download is not set, and format is set, make sure the user knows that
 	// setting format won't do anything
 	if !download && r.FormValue("format") != "" && r.FormValue("format") != "json" {
-		util.WriteErrResponse(w, http.StatusBadRequest, fmt.Errorf("the format must be json if used without the download parameter"))
-		return
+		return nil, fmt.Errorf("the format must be json if used without the download parameter")
 	}
 
 	p := &lib.GetParams{
-		Path:     d.String(),
+		Path:     path,
 		Format:   format,
 		Selector: "body",
 		Limit:    listParams.Limit,
 		Offset:   listParams.Offset,
-		All:      r.FormValue("all") == "true" && listParams.Limit == defaultDataLimit && listParams.Offset == 0,
+		All:      r.FormValue("all") == "true" && !readOnly,
```

Comment:
in a future PR we should deprecate the `All` field from GetParams in favour of directly supplying`Offset: 0, Limit: -1`. This is great in the meantime